internal/api/mfa.go

```diff
@@ -246,12 +323,22 @@ func (a *API) VerifyFactor(w http.ResponseWriter, r *http.Request) error {
 		return internalServerError("Database error verifying MFA TOTP secret").WithInternalError(err)
 	}
 
-	valid, verr := totp.ValidateCustom(params.Code, secret, time.Now().UTC(), totp.ValidateOpts{
-		Period:    30,
-		Skew:      1,
-		Digits:    otp.DigitsSix,
-		Algorithm: otp.AlgorithmSHA1,
-	})
+	var valid bool
+	var verr error
+	var authenticationMethod models.AuthenticationMethod
+	if factor.IsTOTPFactor() {
+		valid, verr = totp.ValidateCustom(params.Code, secret, time.Now().UTC(), totp.ValidateOpts{
+			Period:    30,
+			Skew:      1,
+			Digits:    otp.DigitsSix,
+			Algorithm: otp.AlgorithmSHA1,
+		})
+		if verr != nil {
+			return verr
+		}
+	} else if factor.IsSMSFactor() {
+		valid = challenge.OtpCode == params.Code
```

Comment:
Always always always use time-safe comparison!

Reply:
https://pkg.go.dev/crypto/subtle#ConstantTimeCompare

Reply:
good catch, thanks!

Reply:
Also please use EncryptedString for the OtpCode!